subprojects/core/src/main/java/org/gradle/api/artifacts/ComponentMetadataSupplier.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A component metadata rule is responsible for providing the initial metadata of a component
+ * from a remote repository, in place of parsing the descriptor. Users may implement a provider
+ * to make dependency resolution faster.
+ *
+ * @since 3.5
+ */
+@Incubating
+public interface ComponentMetadataSupplier {
```

Comment:
Right, but why not at least use `ComponentMetadataSupplier extends Action<? super ComponentMetadataSupplierDetails>`, so that this is a bit more consistent with our other dependency resolution rules? My assumption is that we want a common mechanism for users to supply a rule implementation: you're saying "the pattern is different" but in my mind the use-case is the same, we've just implemented a different mechanism. That's fine, but then let's do some work to move toward a common mechanism for all rule implementations.